src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java

```diff
@@ -2,49 +2,74 @@
+import htsjdk.variant.variantcontext.VariantContext;
 import org.apache.beam.sdk.io.TextIO;
 import org.apache.beam.sdk.options.PipelineOptions;
 import org.apache.beam.sdk.Pipeline;
 import org.apache.beam.sdk.transforms.Filter;
 import org.apache.beam.sdk.transforms.MapElements;
 import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
 import org.apache.beam.sdk.values.TypeDescriptors;
 
 public final class VcfToBqPipelineRunner implements PipelineRunner {
 
   private final VcfToBqContext context;
   private final PipelineOptions options;
   private final VcfParser vcfParser;
+  private final BigQueryRowGenerator bigQueryRowGenerator;
+
+  private final TupleTag<TableRow> VALID_VARIANT_TO_BQ_RECORD_TAG = new TupleTag<>() {};
+  private final TupleTag<String> MALFORMED_RECORD_ERROR_MESSAGE_TAG = new TupleTag<>() {};
 
   /** Implementation of {@link PipelineRunner} service. */
   @Inject
   public VcfToBqPipelineRunner(
-      VcfToBqContext context, VcfToBqOptions options, VcfParser vcfParser) {
+      VcfToBqContext context, VcfToBqOptions options, VcfParser vcfParser,
+      BigQueryRowGenerator bigQueryRowGenerator) {
     this.context = context;
     this.options = (PipelineOptions) options;
     this.vcfParser = vcfParser;
+    this.bigQueryRowGenerator = bigQueryRowGenerator;
   }
 
   public void runPipeline() {
-    // Demo code.
     Pipeline pipeline = Pipeline.create(options);
-    pipeline.apply(TextIO.read().from(context.getInputFile()))
-      .apply(Filter.by((String inputLine) -> !inputLine.startsWith("#")))
-      .apply(ParDo.of(new ConvertLineToVariantFn(vcfParser, context.getHeaderLines())))
-      .apply(
-          MapElements
-              .into(TypeDescriptors.strings())
-              .via(
-                  (Variant variant) ->
-                  String.format("Contig: %s; Start: %d; End: %d",
-                      variant.getContig(), variant.getStart(), variant.getEnd())))
-      .apply(TextIO.write().to(context.getOutput()).withNoSpilling());
+    PCollection<VariantContext> variantContextPCollection = pipeline
+        .apply(TextIO.read().from(context.getInputFile()))
+        .apply(Filter.by((String inputLine) -> !inputLine.startsWith("#")))
+        .apply(ParDo.of(new ConvertLineToVariantFn(vcfParser, context.getHeaderLines())));
+    PCollectionTuple tableRowTuple = variantContextPCollection
+        .apply("VariantContextToBQRow",
+            ParDo.of(new ConvertVariantToRowFn(bigQueryRowGenerator,
+                context.getVCFHeader(), context.getAllowMalformedRecords(),
+                    VALID_VARIANT_TO_BQ_RECORD_TAG, MALFORMED_RECORD_ERROR_MESSAGE_TAG))
+                .withOutputTags(VALID_VARIANT_TO_BQ_RECORD_TAG, TupleTagList.of(MALFORMED_RECORD_ERROR_MESSAGE_TAG)));

```

Comment:
Why is second one provided as list, while the first tuple tag is a single element? Is that the requirement?

Also, 100 chars.

Reply:
Yes. It's main tag + list of additional tags